File: apps/api/app/services/publishers/instagram.py
"""Instagram publisher — two-step container → publish, carousel support."""

import asyncio
import logging
from datetime import datetime, timezone

import httpx

from app.services.publishers.base import BasePublisher, PublishResult
from app.utils.exceptions import ValidationError

logger = logging.getLogger(__name__)

GRAPH_API = "https://graph.instagram.com/v21.0"
INSTAGRAM_CAPTION_MAX = 2200
CONTAINER_POLL_INTERVAL = 2
CONTAINER_POLL_MAX = 30


class InstagramPublisher(BasePublisher):
    platform_id = "instagram"

    def validate(self, output: dict) -> None:
        content = output.get("content", "")
        media_urls = output.get("media_urls", [])
        if not media_urls:
            raise ValidationError(
                message="Media required",
                detail="Instagram requires at least one image or video.",
            )
        if len(content) > INSTAGRAM_CAPTION_MAX:
            raise ValidationError(
                message="Caption too long",
                detail=f"Caption is {len(content)} chars (max {INSTAGRAM_CAPTION_MAX}).",
            )

    async def publish(self, output: dict, token: str, **kwargs) -> PublishResult:
        platform_user_id = kwargs.get("platform_user_id", "")
        media_urls = output.get("media_urls", [])
        caption = output.get("content", "")

        if len(media_urls) > 1:
            return await self._publish_carousel(platform_user_id, media_urls, caption, token)
        return await self._publish_single(platform_user_id, media_urls[0], caption, token)

    async def _publish_single(
        self, ig_user_id: str, media_url: str, caption: str, token: str
    ) -> PublishResult:
        """Two-step: create media container, then publish."""
        async with httpx.AsyncClient(timeout=30) as client:
            # Step 1: Create container
            resp = await client.post(
                f"{GRAPH_API}/{ig_user_id}/media",
                params={
                    "image_url": media_url,
                    "caption": caption,
                    "access_token": token,
                },
            )

        if resp.status_code != 200:
            logger.error("Instagram container creation failed: %s", resp.text)
            return PublishResult(
                success=False, platform="instagram",
                error=f"Container creation failed: {resp.status_code}",
            )

        container_id = resp.json().get("id")
        await self._wait_for_container(container_id, token)

        # Step 2: Publish
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.post(
                f"{GRAPH_API}/{ig_user_id}/media_publish",
                params={"creation_id": container_id, "access_token": token},
            )

        if resp.status_code != 200:
            logger.error("Instagram publish failed: %s", resp.text)
            return PublishResult(
                success=False, platform="instagram",
                error=f"Publish failed: {resp.status_code}",
            )

        post_id = resp.json().get("id", "")
        return PublishResult(
            success=True,
            platform="instagram",
            platform_post_id=post_id,
            platform_post_url=f"https://www.instagram.com/p/{post_id}/",
            published_at=datetime.now(timezone.utc),
        )

    async def _publish_carousel(
        self, ig_user_id: str, media_urls: list[str], caption: str, token: str
    ) -> PublishResult:
        """Create multiple containers, then a carousel container, then publish."""
        children_ids = []

        for url in media_urls[:10]:  # Instagram max 10 items
            async with httpx.AsyncClient(timeout=30) as client:
                resp = await client.post(
                    f"{GRAPH_API}/{ig_user_id}/media",
                    params={
                        "image_url": url,
                        "is_carousel_item": "true",
                        "access_token": token,
                    },
                )
            if resp.status_code != 200:
                logger.error("Instagram carousel child failed: %s", resp.text)
                return PublishResult(
                    success=False, platform="instagram",
                    error=f"Carousel child creation failed: {resp.status_code}",
                )
            children_ids.append(resp.json().get("id"))

        # Create carousel container
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.post(
                f"{GRAPH_API}/{ig_user_id}/media",
                params={
                    "media_type": "CAROUSEL",
                    "caption": caption,
                    "children": ",".join(children_ids),
                    "access_token": token,
                },
            )

        if resp.status_code != 200:
            logger.error("Instagram carousel container failed: %s", resp.text)
            return PublishResult(
                success=False, platform="instagram",
                error=f"Carousel container failed: {resp.status_code}",
            )

        container_id = resp.json().get("id")
        await self._wait_for_container(container_id, token)

        # Publish carousel
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.post(
                f"{GRAPH_API}/{ig_user_id}/media_publish",
                params={"creation_id": container_id, "access_token": token},
            )

        if resp.status_code != 200:
            return PublishResult(
                success=False, platform="instagram",
                error=f"Carousel publish failed: {resp.status_code}",
            )

        post_id = resp.json().get("id", "")
        return PublishResult(
            success=True,
            platform="instagram",
            platform_post_id=post_id,
            platform_post_url=f"https://www.instagram.com/p/{post_id}/",
            published_at=datetime.now(timezone.utc),
            metadata={"carousel_items": len(children_ids)},
        )

    async def _wait_for_container(self, container_id: str, token: str) -> None:
        """Poll container status until FINISHED or timeout."""
        for _ in range(CONTAINER_POLL_MAX):
            async with httpx.AsyncClient(timeout=10) as client:
                resp = await client.get(
                    f"{GRAPH_API}/{container_id}",
                    params={"fields": "status_code", "access_token": token},
                )
            if resp.status_code == 200:
                status = resp.json().get("status_code")
                if status == "FINISHED":
                    return
                if status == "ERROR":
                    raise ValidationError(
                        message="Container error",
                        detail="Instagram media container processing failed.",
                    )
            await asyncio.sleep(CONTAINER_POLL_INTERVAL)

File: functions/src/lib/platforms/publishers/instagram.ts
/**
 * Instagram Graph API publisher — container-based publish flow.
 * Uses Meta Graph API: create container → poll status → publish.
 */

import type { PlatformPublisher, PublishResult, DecryptedTokens, RefreshedTokens } from "./base.js";

const GRAPH_URL = "https://graph.instagram.com/v21.0";
const GRAPH_OAUTH_URL = "https://graph.instagram.com/oauth/access_token";

export class InstagramPublisher implements PlatformPublisher {
  async publish(content: string, metadata: Record<string, unknown>, tokens: DecryptedTokens): Promise<PublishResult> {
    const userId = tokens.platformUserId;
    if (!userId) {
      return { success: false, postId: null, url: null, error: "No Instagram user ID found. Reconnect your account." };
    }

    const formatType = (metadata.format_type as string) || "instagram_caption";

    try {
      if (formatType === "instagram_carousel" && metadata.image_urls) {
        return await this.publishCarousel(content, metadata, tokens.accessToken, userId);
      }
      return await this.publishSingleCaption(content, metadata, tokens.accessToken, userId);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return { success: false, postId: null, url: null, error: `Instagram publish error: ${message}` };
    }
  }

  private async publishSingleCaption(
    content: string,
    metadata: Record<string, unknown>,
    accessToken: string,
    userId: string,
  ): Promise<PublishResult> {
    const imageUrl = (metadata.image_url as string) || "";

    // Step 1: Create media container
    const containerParams: Record<string, string> = {
      caption: content,
      access_token: accessToken,
    };

    if (imageUrl) {
      containerParams.image_url = imageUrl;
    } else {
      // Text-only posts require an image on Instagram — return guidance
      return {
        success: false,
        postId: null,
        url: null,
        error: "Instagram requires an image. Attach an image_url in metadata to publish.",
      };
    }

    const createResp = await fetch(`${GRAPH_URL}/${userId}/media`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams(containerParams),
    });

    if (!createResp.ok) {
      const body = await createResp.text();
      throw new Error(`Container creation failed ${createResp.status}: ${body}`);
    }

    const containerData = await createResp.json() as Record<string, unknown>;
    const containerId = containerData.id as string;

    // Step 2: Wait for container to be ready (poll status)
    await this.waitForContainer(containerId, accessToken);

    // Step 3: Publish the container
    const publishResp = await fetch(`${GRAPH_URL}/${userId}/media_publish`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        creation_id: containerId,
        access_token: accessToken,
      }),
    });

    if (!publishResp.ok) {
      const body = await publishResp.text();
      throw new Error(`Publish failed ${publishResp.status}: ${body}`);
    }

    const publishData = await publishResp.json() as Record<string, unknown>;
    const postId = publishData.id as string;

    return {
      success: true,
      postId,
      url: `https://www.instagram.com/p/${postId}/`,
      error: null,
    };
  }

  private async publishCarousel(
    content: string,
    metadata: Record<string, unknown>,
    accessToken: string,
    userId: string,
  ): Promise<PublishResult> {
    const imageUrls = (metadata.image_urls as string[]) || [];
    if (imageUrls.length < 2) {
      return this.publishSingleCaption(content, { ...metadata, image_url: imageUrls[0] }, accessToken, userId);
    }

    // Step 1: Create child containers for each image
    const childIds: string[] = [];
    for (const imageUrl of imageUrls.slice(0, 10)) {
      const resp = await fetch(`${GRAPH_URL}/${userId}/media`, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: new URLSearchParams({
          image_url: imageUrl,
          is_carousel_item: "true",
          access_token: accessToken,
        }),
      });

      if (!resp.ok) {
        const body = await resp.text();
        throw new Error(`Carousel child creation failed ${resp.status}: ${body}`);
      }

      const data = await resp.json() as Record<string, unknown>;
      childIds.push(data.id as string);
    }

    // Wait for all children to process
    for (const childId of childIds) {
      await this.waitForContainer(childId, accessToken);
    }

    // Step 2: Create carousel container
    const carouselResp = await fetch(`${GRAPH_URL}/${userId}/media`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        caption: content,
        media_type: "CAROUSEL",
        children: childIds.join(","),
        access_token: accessToken,
      }),
    });

    if (!carouselResp.ok) {
      const body = await carouselResp.text();
      throw new Error(`Carousel container creation failed ${carouselResp.status}: ${body}`);
    }

    const carouselData = await carouselResp.json() as Record<string, unknown>;
    const carouselId = carouselData.id as string;

    await this.waitForContainer(carouselId, accessToken);

    // Step 3: Publish carousel
    const publishResp = await fetch(`${GRAPH_URL}/${userId}/media_publish`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        creation_id: carouselId,
        access_token: accessToken,
      }),
    });

    if (!publishResp.ok) {
      const body = await publishResp.text();
      throw new Error(`Carousel publish failed ${publishResp.status}: ${body}`);
    }

    const publishData = await publishResp.json() as Record<string, unknown>;
    const postId = publishData.id as string;

    return {
      success: true,
      postId,
      url: `https://www.instagram.com/p/${postId}/`,
      error: null,
    };
  }

  private async waitForContainer(containerId: string, accessToken: string, maxAttempts = 10): Promise<void> {
    for (let i = 0; i < maxAttempts; i++) {
      const resp = await fetch(
        `${GRAPH_URL}/${containerId}?fields=status_code&access_token=${accessToken}`,
      );

      if (resp.ok) {
        const data = await resp.json() as Record<string, unknown>;
        if (data.status_code === "FINISHED") return;
        if (data.status_code === "ERROR") {
          throw new Error("Instagram media container processing failed.");
        }
      }

      await new Promise((resolve) => setTimeout(resolve, 2000));
    }
    throw new Error("Instagram media container timed out waiting for processing.");
  }

  async validateConnection(_tokens: DecryptedTokens, expiresAt: Date | null): Promise<boolean> {
    if (expiresAt) return expiresAt > new Date();
    return true;
  }

  async refreshToken(tokens: DecryptedTokens): Promise<RefreshedTokens | null> {
    // Instagram long-lived tokens can be refreshed via GET endpoint
    if (!tokens.accessToken) return null;

    const resp = await fetch(
      `${GRAPH_URL}/refresh_access_token?grant_type=ig_refresh_token&access_token=${tokens.accessToken}`,
    );

    if (!resp.ok) return null;

    const data = await resp.json() as Record<string, unknown>;
    return {
      accessToken: data.access_token as string,
      refreshToken: null,
      expiresIn: (data.expires_in as number) ?? 5184000, // 60 days default
    };
  }
}
